controlplane/internal/controllers/rke2controlplane_controller.go

```diff
@@ -339,11 +339,13 @@ func (r *RKE2ControlPlaneReconciler) updateStatus(ctx context.Context, rcp *cont
 	}
 
 	if len(ownedMachines) == 0 {
-		return fmt.Errorf("no Control Plane Machines exist for RKE2ControlPlane %s/%s", rcp.Namespace, rcp.Name)
+		r.Log.Info(fmt.Sprintf("no Control Plane Machines exist for RKE2ControlPlane %s/%s", rcp.Namespace, rcp.Name))
+		return nil
 	}
 
 	if len(readyMachines) == 0 {
-		return fmt.Errorf("no Control Plane Machines are ready for RKE2ControlPlane %s/%s", rcp.Namespace, rcp.Name)
+		r.Log.Info(fmt.Sprintf("no Control Plane Machines are ready for RKE2ControlPlane %s/%s", rcp.Namespace, rcp.Name))
+		return nil
```

Comment:
This changed, combined with the previous change, will cause a wrongly set `rcp.Status.Ready`.
Indeed, when you continue the logic without triggering a `Requeue`, you will find this:

```go
if len(readyMachines) == len(ownedMachines) {
	rcp.Status.Ready = true
}
```
If both `ownedMachines` and `readyMachines` collections are empty, the if condition will be true, and the `rcp.Status.Ready` will be `true` : this is not the desired behaviour. We need to keep requeuing whenever `ownedMachines` or `readyMachines` is empty.

Reply:
If either one of `ownedMachines` or `readyMachines` is empty we still return, what this patch does is return with a `nil` error, so we won't hit the if condition you mention. Requeing was not happening before this patch, if either one of collection is empty we'd return an error from `updateStatus` that gets assigned to `reterr`:
```
		if err := r.updateStatus(ctx, rcp, cluster); err != nil {
			var connFailure *rke2.RemoteClusterConnectionError
			if errors.As(err, &connFailure) {
				logger.Info("Could not connect to workload cluster to fetch status", "err", err.Error())
			} else {
				logger.Error(err, "Failed to update RKE2ControlPlane Status")
				reterr = kerrors.NewAggregate([]error{reterr, err})
			}
		}
```

Due to `reterr != nil` we won't requeue (with the default time) thus leading to high times of control plane readiness:

```
		if reterr == nil && !res.Requeue && res.RequeueAfter <= 0 && rcp.ObjectMeta.DeletionTimestamp.IsZero() {
			if !rcp.Status.Ready {
				res = ctrl.Result{RequeueAfter: DefaultRequeueTime}
			}
		}
```

Reply:
You are right in terms of behaviour, I didn't read well the return stuff.
Now, I am wondering in terms of code readability and meaning. Is it better to:
- return an error and handle it in the `defer` func after the` r.updateStatus` call
- return nil and keep the `defer `func as it is

And I do lean into the second one (your approach) because not having ready machines is not an actual failure, it just needs a requeue.